worker/undertaker/manifold.go

```diff
@@ -69,3 +75,18 @@ func Manifold(config ManifoldConfig) dependency.Manifold {
 		Start: config.start,
 	}
 }
+
+func isErrMissing(err error) bool {
+	return errors.Cause(err) == dependency.ErrMissing
+}
+
+// unavailableDestroyer is an environs.CloudDestroyer that always
+// fails to destroy. We use it when the real environ isn't available
+// because the cloud credentials are invalid so that the undertaker
+// can still remove the model if destruction is forced.
+type unavailableDestroyer struct{}
+
+// Destroy is part of environs.CloudDestroyer.
+func (*unavailableDestroyer) Destroy(ctx envcontext.ProviderCallContext) error {
+	return errors.New("cloud environment unavailable")
```

Comment:
I am not sure that we want to return an error here...

At the very least, I would have expected to a change on the worker side - how do we handle the error in the actual undertaker worker? Currently, we'd err out in the run() ... But since it is valid for us now to have an undertaker that does not communicate with the cloud, maybe u.config.Destroyer.Destroy() call should not err?

Reply:
#10245 changed the worker to log the error and continue if `destroy-model --force` is run.

If we don't return an error here then destroy-model will work without force when the credential is invalid, which we definitely don't want - that could lead to people not realising they've got instances still running in the cloud. As long as we make them specify `--force` then they (should be) aware that things might be left.